### PR TITLE
Core: Increase address space limits and rework Windows address space initialization.

### PR DESCRIPTION
**Pull request by StevenMiller123**
Current Windows builds of shadPS4 are prone to asserting in address_space on emulator initialization due to Windows' address space randomization. This also prevents expanding the address space to higher addresses (which is needed for a handful of EA titles).

This PR introduces logic to detect and map around system mappings on Windows devices, pass information about these gaps to our memory code, and more robust address validation logic to ensure gaps are accounted for when detecting if requested memory areas are within the vma map.

For Mac users: this PR should make asserts trigger more consistently when games try mapping to addresses that we can't reserve.

For Windows users: this PR should fix those random-ish cases of
```
[Debug] <Critical> address_space.cpp:105 operator(): Assertion Failed!
Unable to reserve virtual address space: Not enough memory resources are available to process this command.
```

For all users: this PR should address cases of the following assert triggering on addresses greater than 0x11000000000 (mainly seen in earlier titles published by EA).

```
[Debug] memory.cpp:902 operator(): Assertion Failed!
Input address .... is out of bounds
```
I'm opening this as a draft for now since I need to test edge cases further, and ideally make sure our own HLE mappings aren't going to hit asserts in the more extreme cases.

I would also appreciate some tests from the aforementioned EA games, the only thing I have hitting that out of bounds assert is an app, and it dies from a Windows-specific issue with code patches.

### CODE REVIEW

**Comment by Missake212** [comment]
EA SPORTS™ UFC® doesn't crash on the memory assert anymore.

[CUSA00264.log](https://github.com/user-attachments/files/22678369/CUSA00264.log)

File: src/core/memory.cpp [hunk 1]
```diff
@@ -138,9 +132,9 @@ void MemoryManager::CopySparseMemory(VAddr virtual_addr, u8* dest, u64 size) {
 }
 
 bool MemoryManager::TryWriteBacking(void* address, const void* data, u32 num_bytes) {
-    ASSERT_MSG(IsValidAddress(address), "Attempted to access invalid address {}",
-               fmt::ptr(address));
     const VAddr virtual_addr = std::bit_cast<VAddr>(address);
+    ASSERT_MSG(IsValidMapping(virtual_addr, num_bytes), "Attempted to access invalid address {}",
```
Comment (Exhigh) [inline src/core/memory.cpp]:
```suggestion
    ASSERT_MSG(IsValidMapping(virtual_addr, num_bytes), "Attempted to access invalid address {:#x}",
```

**Comment by StevenMiller123** [comment]
I've abused AMD's Windows drivers to test under extreme circumstances and everything seems to work as expected there (most of the system reserved area and a chunk of the user area is completely inaccessible if those drivers initialize before us).
Games that try performing fixed mappings in that inaccessible space hit asserts, but that's intended behavior in this PR.

Since it seems to work fine for me, I'm marking this as ready for review.

**Comment by StevenMiller123** [comment]
Linux is not a fan of the increased user size, so marking as draft while I figure this out.

**Comment by StevenMiller123** [comment]
Latest push is just to fix Linux on my end. From what I've seen in logs, nothing should regress from this change.

**Comment by StevenMiller123** [comment]
Seems like Mac doesn't behave properly with the increased user size. Leaving this as a draft until that can be addressed.

**Comment by StevenMiller123** [comment]
Windows 10 devices seem to completely hang with this PR. Until I've fixed that issue, I suggest Windows 10 users avoid testing this.

**Comment by StevenMiller123** [comment]
The Windows issue comes down to either dropping Windows 10 support entirely or breaking a couple EA titles. Needless to say, I've chosen the latter option.

I need Windows 10 testers to test for performance issues, that may still end up a problem with this current push.

**Comment by StevenMiller123** [comment]
UFC 1 should still work, hopefully.

**Comment by StevenMiller123** [comment]
This probably also fixes any regressions on Mac OS too, I believe this level of address space expanding was tested on there during #2872

**Comment by StevenMiller123** [comment]
I've reverted the "fun" part of this PR for now due to limitations with Windows 10.

The improved logic will still help for Windows users hitting those random asserts on emulator startup, and make asserts more consistent for Mac users. Games like UFC will be back to their compatibility on main though.

**Comment by StevenMiller123** [comment]
With the help of @squidbus, the high user max should be working for Mac OS devices now.
I've also added checks to allow high user max for Windows 11 devices, and prevent it on Windows 10 devices where VirtualAlloc2 is prone to hanging.

I still need to take a better look at the issues with higher user max on Linux, so this will remain a draft for now. I also need someone to test Windows 10 behavior.

File: src/core/libraries/gnmdriver/gnmdriver.cpp [hunk 2]
```diff
@@ -70,7 +71,12 @@ static bool send_init_packet{true}; // initialize HW state before first game's s
 static int sdk_version{0};
 
 static u32 asc_next_offs_dw[Liverpool::NumComputeRings];
-static constexpr VAddr tessellation_factors_ring_addr = Core::SYSTEM_RESERVED_MAX - 0xFFFFFFF;
+
+// We need to get an instance of our address space to derive a proper ring address here.
+auto* memory = Core::Memory::Instance();
+auto& address_space = memory->GetAddressSpace();
+static VAddr tessellation_factors_ring_addr = address_space.SystemReservedVirtualBase() +
+                                              address_space.SystemReservedVirtualSize() - 0xFFFFFFF;
```
Comment (squidbus) [inline src/core/libraries/gnmdriver/gnmdriver.cpp]: Not sure if it's safe to initialize this statically, may be safer to initialize this on first `sceGnmGetTheTessellationFactorRingBufferBaseAddress` call.

**Comment by StevenMiller123** [comment]
With the newest changes, all OSes now load the shadPS4 executable at address `0x700000000000`, which allows Linux to use the full USER_MAX we've defined, and helps reserve some extra lower addresses on the Windows side.

This needs testing on Windows 10 though, I don't know if the same VirtualAlloc2 performance issue will end up applying to this too.

**Comment by StevenMiller123** [comment]
Local tests suggest Windows 10 works fine, and that my Windows 10 detection code is functioning.
Just need to fix Linux actions now

**Comment by StevenMiller123** [comment]
I've stuck to just reducing USER_MAX on Linux for now, this shouldn't impact too much since it's still high enough to include the rather extreme fixed mappings EA's titles perform.

Like with the Windows 10 issue, I'm open to changing this if anyone's able to suggest a working solution.

**Comment by StevenMiller123** [comment]
Everything seems to be working properly on my end, so I'm marking this as ready for review.

**Comment by StevenMiller123** [comment]
Given the variety of games hitting this particular memory assert, here's a list of games that this PR should affect based on compatibility reports:
- Need for Speed™
- FIFA 14
- NHL™ 15
- Madden NFL 16
- FIFA 16
- STAR WARS™ Battlefront™ - Only version 1.00 was affected.
- EA SPORTS™ UFC® - Confirmed improvements.

None of these will improve on Windows 10, since I've had to restrict the user size to accomodate OS bugs there. Windows 11, Mac OS, and Linux should all see improvements though.

This PR should also improve similar Windows-specific issues (for both Windows 10 and Windows 11) in:
- Saints Row: Gat out of Hell
- Saints Row IV: Re-Elected

These two games were having issues due to the old Windows address space logic, my tests suggest the addresses they're trying to use should be in the vma_map with this PR now.

**Comment by Missake212** [comment]
Can confirm EA SPORTS™ UFC® status is still improved by this PR.

File: src/core/libraries/gnmdriver/gnmdriver.cpp [hunk 3]
```diff
@@ -996,7 +999,14 @@ int PS4_SYSV_ABI sceGnmGetShaderStatus() {
 }
 
 VAddr PS4_SYSV_ABI sceGnmGetTheTessellationFactorRingBufferBaseAddress() {
+    auto* memory = Core::Memory::Instance();
+    auto& address_space = memory->GetAddressSpace();
```
Comment (squidbus) [inline src/core/libraries/gnmdriver/gnmdriver.cpp]: Can move these into the `if` case.

**Review by squidbus — COMMENTED** [review]
Just one small comment

**Review by squidbus — APPROVED** [review]
